feat: Jump to last message WPB-6518

svitovyda opened the pull request:
<!--do not remove this marker, its needed to replace info when ticket title is updated -->
<!--jira-description-action-hidden-marker-start-->

<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-6518" title="WPB-6518" target="_blank"><img alt="Story" src="https://wearezeta.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10815?size=medium" />WPB-6518</a>  [Web] Jump to last message in a conversation
  </td></table>
  <br />
 

<!--jira-description-action-hidden-marker-end-->
<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

## Description

By the product/design team request, the user should jump to the very last message in the conversation, marking all the messages as read

## Screenshots/Screencast (for UI changes)


https://github.com/wireapp/wire-webapp/assets/24770159/03c63fba-468f-4f39-ac81-0f0836f8c619


https://github.com/wireapp/wire-webapp/assets/24770159/38e094ca-8af9-4c10-a661-971d0bfd9ef5


https://github.com/wireapp/wire-webapp/assets/24770159/be9d6045-6103-4c6f-a3d5-7f2333121007


## Checklist

- [x] PR has been self reviewed by the author;
- [x] Hard-to-understand areas of the code have been commented;
- [x] If it is a core feature, unit tests have been added;

Commit:
feat: Add "Jump to las message" button WPB-6518

Commit:
Show/hide "jump to last message" button on visibility of the last message;
Mark all the conversation as read on the "jump to last message" button click;
Reload the conversation on the "jump to last message" click if the latest conversation messages were not load;

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/wireapp/wire-webapp?pullRequest=17386) <br/>All committers have signed the CLA.

svitovyda commented on src/script/components/LastMessageTracker/LastMessageTracker.tsx:
```diff
@@ -0,0 +1,79 @@
+  isVisible: boolean;
+}
+
+export interface LastMessageTrackerProps extends HTMLProps<HTMLElement> {
```
This should be a separate component as we want to avoid the entire content of `Conversation` to re-render each time messages are scrolled and messages visibility changes

svitovyda commented on src/script/components/MessagesList/Message/index.tsx:
```diff
@@ -88,6 +89,7 @@ export const Message: React.FC<MessageParams & {scrollTo?: ScrollToElement}> = p
     isHighlighted,
     hideHeader,
     onVisible,
+    onVisibilityChange,
```
I know this is weird, but as `isVisible` was implemented till now, it would be triggered only the first time (and if it's added as the item of the list one by one, and then disappear from the visible area - it will still be called). I don't mind merging them at least on the `Message` and the level of its usage, but I want to hear your opinions on how to do it the best

Commit:
Fix the way last message visibility is tracked;
Add re-render of message list on click on button when last message is loaded;

V-Gira committed:
runfix: add breakpoint to bottom positioning (#17483)

Commit:
Fix the way conversation is marked as read;
Make isLastReceivedMessage independent function and use it for last message visibility;
Fix tests;

svitovyda commented on src/script/components/Conversation/Conversation.tsx:
```diff
@@ -460,6 +464,19 @@ export const Conversation = ({
 
+  const onGoToLastMessage = () => {
+    if (activeConversation) {
+      activeConversation.initialMessage(undefined);
```
clean up anything like search result

atomrc replied:
This would be a valuable comment to add to the code directly

svitovyda replied:
done

svitovyda commented on src/script/components/MessagesList/MessageList.tsx:
```diff
@@ -168,7 +166,7 @@ export const MessagesList: FC<MessagesListParams> = ({
   }, [messagesContainer?.parentElement, loaded, filteredMessages, selfUser?.id]);
 
-  // Listen to resizes of the the content element (if it's resized it means something has changed in the message list, link a link preview was generated)
+  // Listen to resizes of the content element (if it's resized it means something has changed in the message list, link a link preview was generated)
```
just fix a typo

svitovyda commented on src/script/components/MessagesList/MessageList.tsx:
```diff
@@ -126,7 +124,7 @@ export const MessagesList: FC<MessagesListParams> = ({
   const groupedMessages = groupMessagesBySenderAndTime(filteredMessages, conversationLastReadTimestamp.current);
 
-  const [messagesContainer, setMessageContainer] = useState<HTMLDivElement | null>(null);
+  const [messagesContainer, setMessagesContainer] = useState<HTMLDivElement | null>(null);
```
just fix a typo

svitovyda commented on setupTests.js:
```diff
@@ -68,5 +68,12 @@ window.z = {userPermission: {}};
 window.URL.createObjectURL = jest.fn();
 window.URL.revokeObjectURL = jest.fn();
 
+Object.defineProperty(document, 'elementFromPoint', {
```
I fixed it globally for tests like `MessageList`, where we don't want to test this functionality, but it gets called now. It is still possible to mock it like `document.elementFromPoint = () => element;`

svitovyda commented on src/script/components/Conversation/Conversation.tsx:
```diff
@@ -460,6 +464,19 @@ export const Conversation = ({
+  const onGoToLastMessage = () => {
+    if (activeConversation) {
+      activeConversation.initialMessage(undefined);
+      if (!activeConversation.hasLastReceivedMessageLoaded()) {
```
there are two scenarios: all messages are loaded and we only need to scroll down - then we call the `rerenderMessageList()` and MessageList gets re-rendered (unread messages will be automatically marked as read). Or we have unloaded messages in the conversation and need to reload it entirely and mark all messages as read to enforce jumping to the very last message instead of the first unread one

atomrc replied:
This is also valuable code comment :)

svitovyda replied:
done

aweiss-dev commented on src/script/view_model/ContentViewModel.ts:
```diff
@@ -239,7 +239,7 @@ export class ContentViewModel {
       openFirstSelfMention = false,
       openNotificationSettings = false,
-    } = options;
+    } = options || {};
```
Options are not supposed to be null | undefined, what is this change for?

svitovyda replied:
Our TS config allowed me to use this function without providing this parameter and I got an exception. And actually, in many cases we provide just `{}`, so I'd rather make `options` optional and remove those `{}`, WDYT?

aweiss-dev replied:
That would be the preferred way to act on this issue. If it can be undefined, it needs to be typed as such 👍

svitovyda replied:
done

atomrc commented on src/script/components/Conversation/Conversation.tsx:
```diff
@@ -118,8 +121,11 @@ export const Conversation = ({
 
   const [isMsgElementsFocusable, setMsgElementsFocusable] = useState(true);
 
+  const [messagesListRerenderKey, rerenderMessageList] = useComponentRerenderKey('messages-list');
```
This seems like a non-trivial hook. Can you add comment to explain why this is needed and what's the purpose of it?

svitovyda replied:
This is the way to trigger the re-render of `MessageList`, by changing its `key` (KUDOs to @PatrykBuniX btw). Should I add a comment to the code?

atomrc replied:
That's clever solution actually. I like it!!

But, because it's clever, it does call for a code-comment :)

svitovyda replied:
done

atomrc commented on src/script/components/Conversation/Conversation.tsx:
```diff
@@ -460,6 +464,19 @@ export const Conversation = ({
     [addReadReceiptToBatch, repositories.conversation, repositories.integration, updateConversationLastRead],
   );
 
+  const onGoToLastMessage = () => {
```
This is a concrete implementation of a handler. The name should not start with `on` it should have a imperative naming (explaining what it does exactly). 

I'd suggest `loadLastReceivedMessages` or something along those lines

svitovyda replied:
done, I called it `jumpToLastMessage`

atomrc commented on src/script/components/Conversation/Conversation.tsx:
```diff
@@ -566,3 +596,26 @@ export const Conversation = ({
+      <ChevronIcon css={{rotate: '90deg', height: 16, width: 16, path: {fill: '#0667C8'}}} />
+    </IconButton>
+  );
+};
```
Let's extract that to its own file

svitovyda replied:
Initially I did so, but then I saw we have such small components in the same file, will do so

svitovyda replied:
done

atomrc commented on src/script/entity/Conversation.ts:
```diff
@@ -163,6 +167,7 @@ export class Conversation {
   public readonly receiptMode: ko.Observable<RECEIPT_MODE>;
   public readonly removed_from_conversation: ko.PureComputed<boolean>;
   public readonly roles: ko.Observable<Record<string, string>>;
+  public readonly isLastMessageVisible: ko.Observable<boolean>;
```
This should not belong in the low level entity. Should be a view-only thing. 
It should leave in `Conversation.tsx` only, I believe

svitovyda replied:
It was my idea too, but then we should pass it to the `MessageList` to implement its updating, or pull all the updating logic into the `Conversation`, chaining props of `onVisible` and `onVisibilityLost`. It was quite ugly implementation. I also thought that some of the props and methods of `ConversationEntity` might not be in the perfect place (like `initialMessage`), and maybe we need a `ConversationView` service, WDYT?

atomrc replied:
I believe we do need a cleaner architecture for those values, yes. But maybe it's not the correct PR to tackle this. Probably a follow up PR would be nice.

Commit:
Move JumpToLastMessageButton into a separate file

atomrc commented on src/script/components/Conversation/Conversation.tsx:
```diff
@@ -460,6 +465,19 @@ export const Conversation = ({
+        activeConversation.release();
+        amplify.publish(WebAppEvents.CONVERSATION.SHOW, activeConversation, {});
+      } else {
+        rerenderMessageList();
```
I think in this case we should just run a scroll animation and not reload the entire conversation. This would feel more natural

Commit:
Add comment for a non-trivial usage of `key` to enforce component re-render;
Add comments for jump to the last message function;
Make options optional in `showConversation`;
Fix CSS for the button and disable eslint for the WithConditionalCSSProp impert;

svitovyda commented on src/script/components/Conversation/JumpToLastMessageButton.tsx:
```diff
@@ -0,0 +1,57 @@
+
+import {FC, HTMLProps} from 'react';
+
+// eslint-disable-next-line import/no-unresolved
```
without this type, there will be test failure due to:
```
   Error: src/script/components/Conversation/JumpToLastMessageButton.tsx(45,6): error TS2322: Type '{ children: Element; accept?: string; acceptCharset?: string; action?: string; allowFullScreen?: boolean; allowTransparency?: boolean; alt?: string; as?: string; async?: boolean; autoComplete?: string; ... 358 more ...; css: { ...; }; }' is not assignable to type 'IconButtonProps<HTMLButtonElement>'.
  Types of property 'ref' are incompatible.
    Type 'LegacyRef<HTMLElement>' is not assignable to type 'Ref<HTMLButtonElement>'.
      Type 'string' is not assignable to type 'Ref<HTMLButtonElement>'.
```

to fix this ESLint issue, StackOverflow recommends to add 
```
extends: [
    "plugin:import/errors",
    "plugin:import/warnings",
    "plugin:import/typescript",
  ],
```
but it doesn't help (and we already have some of them in the `@wireapp/eslint-config`

Without using `{...rest}` the CSS will not get applied on `IconButton`, and setting the position CSS inside of the `JumpToLastMessageButton` looks as a bad solution, it should be positioned where it is used

tlebon replied:
seems like a typing issue

tlebon replied:
i would prefer the css inside of the component and just pass the variable in and decide the size based on var

svitovyda replied:
yes, it is a type issue, and using `WithConditionalCSSProp` solves it. but this import causes the `ESLint` error

this is the CSS positioning we have to set:
```
  position: 'absolute',
  bottom: mdBreakpoint? '100px' : '56px',
  right: '10px',
  height: '40px',
```
Do you mean you'd prefer to have a separate custom `CSSObject` prop in `JumpToLastMessageButtonProps` instead of `css`  and apply it on `IconButton`?

svitovyda replied:
we decided to leave this component inside of the `MessageList` file, as it will only be used there, so that we can hardcode all the styles inside of it without any `...rest` or other props way logic

przemvs replied:
I think we don't need to pass css from parent component, so I think better to move this component outside MessageList.tsx :)

svitovyda replied:
done

przemvs commented on src/script/components/Conversation/Conversation.tsx:
```diff
@@ -526,11 +549,22 @@ export const Conversation = ({
+              bottom: mdBreakpoint ? '100px' : '56px',
+              right: '10px',
+              height: '40px',
+            }}
```
Could You move this styles to conversation.style.ts file? :)

svitovyda replied:
done

Commit:
Move jump to last message into MessageList;
Make JumpToLastMessageButton part of MessageList file to not let it be used outside;
Use debounce to avoid too often button visibility change;
Make onVisibilityLost call on component unmount configurable;

Commit:
Fix behavior for the corner case of search result being in the last messages batch

svitovyda commented on src/script/components/MessagesList/MessageList.tsx:
```diff
@@ -245,78 +255,167 @@ export const MessagesList: FC<MessagesListParams> = ({
+  const jumpToLastMessage = () => {
+    if (conversation) {
+      // clean up anything like search result
+      setHighlightedMessage(undefined);
```
It feels that this cleanup should happen somewhere in the `useEffect`, when props and observables change, but I'm afraid to break something this way

svitovyda commented on src/script/components/MessagesList/MessageList.tsx:
```diff
@@ -245,78 +255,167 @@ export const MessagesList: FC<MessagesListParams> = ({
+  );
+};
+
+export interface JumpToLastMessageButtonProps extends HTMLProps<HTMLElement> {
```
we agreed to keep this component inside of `MessageList.tsx` as it contains positioning CSS and we didn't want to overcomplicate it by passing the custom CSS. It still should be a separate component with it's own state, as otherwise `isLastMessageVisible` will make the entire `MessageList` to re-render, unless we will implement memoisation of each element rendered.

svitovyda replied:
`export` is for the test

przemvs replied:
I think You can move it outside MessageList, and use - memo(Component) with validating props :)

svitovyda replied:
there were quite some issues with passing the `css`, we moved it back and forth a few times already. I'd leave it as is for now

svitovyda replied:
done

svitovyda commented on src/script/components/MessagesList/MessageList.tsx:
```diff
@@ -245,78 +255,167 @@ export const MessagesList: FC<MessagesListParams> = ({
+  const [isLastMessageVisible, setIsLastMessageVisible] = useState(conversation.isLastMessageVisible());
+
+  useEffect(() => {
+    const subscription = conversation.isLastMessageVisible.subscribe(
```
I can't use here `useKoSubscribableChildren` as I need to add debouncing to the change of the state.

przemvs replied:
You can do debounce with useCallback component, then You can use useKoSubscribableChildren :)

przemvs replied:
You can do this smth like that:

```
  const {isLastMessageVisible} = useKoSubscribableChildren(conversation, ['isLastMessageVisible']);

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const debounced = useCallback(
    debounce((value: boolean) => value, DEBOUNCE_TIME),
    [],
  );

  if (!debounced(isLastMessageVisible)) {
    return null;
  }

  return ...
```

svitovyda replied:
I tried it, but there is also one more thing - the initial value of `isLastMessageVisible` should be set by the last fired value. Just waiting for the new event and debouncing makes it behave really weird. I suggest to keep my existing implementation

przemvs replied:
Checked, and for now we keep Your code ;)

svitovyda commented on src/script/components/MessagesList/MessageList.tsx:
```diff
@@ -245,78 +255,167 @@ export const MessagesList: FC<MessagesListParams> = ({
+
+  useEffect(() => {
+    const subscription = conversation.isLastMessageVisible.subscribe(
+      debounce(value => {
```
debouncing is needed for messages, sent by `self`: at first the message from self gets mounted as not yet delivered, then it gets unmounted and mounted as received from the server. this makes the jump to the last message button blink. there is a complicated way to check if a message that is unmounted (its `onVisibilityLost` was called) has not yet been delivered, but we should somehow distinguish whether it was unmounted and not hidden by scrolling. I think debouncing is better, while it does add a bit of noticeable delay even with 200 ms

svitovyda commented on src/script/components/utils/InViewport.tsx:
```diff
@@ -25,6 +25,7 @@ import {viewportObserver} from 'Util/DOM/viewportObserver';
 interface InViewportParams {
   onVisible: () => void;
```
in the current implementation, this callback will be called when most of the element is visible. and we need it this way to mark the message as read. if I reuse this callback, the button "jump to the last message" will be disappearing slightly different from how it behaves on our android app - not when a bit of the last message appears, but when the big chunk of it is visible. @atomrc suggested we leave it as it is. the other option would be to implement one callback `onPartialVisibilityChange(visible: boolean)` instead of `onVisibilityLost()` and call it with `true` differently than `onVisible` (as `onVisible` is used in many places, not only for messages)

Commit:
Add test for JumpToLastMessageButton, fix MessageList test

Commit:
fix: Avoid listening to initialMessage change.

Initial message is a static property that is read only when a conversation is loaded. Once loaded this value should not be read again and should not trigger re-rendering

Commit:
fixup! fix: Avoid listening to initialMessage change.

Commit:
Final fix of jump to last message animation

Commit:
Extract styles into separate file

przemvs commented on src/script/components/MessagesList/MessageList.styles.ts:
```diff
@@ -0,0 +1,28 @@
+
+export const jumpToLastMessageButtonStyles: (mdBreakpoint: boolean) => CSSObject = (mdBreakpoint: boolean) => ({
+  position: 'absolute',
+  bottom: mdBreakpoint ? '100px' : '56px',
```
You can do smth like that:

```suggestion
export const jumpToLastMessageButtonStyles: (mdBreakpoint: boolean) => CSSObject = (mdBreakpoint: boolean) => ({
  position: 'absolute',
  bottom: '56px',
  '@media (max-width: @screen-md-min)': {
     bottom: '100px',
  }
```

svitovyda replied:
Looks like media queries don't accept less variables, so I just hardcoded 768px

przemvs commented on src/script/components/MessagesList/MessageList.tsx:
```diff
@@ -245,78 +252,158 @@ export const MessagesList: FC<MessagesListParams> = ({
+  conversation: Conversation;
+}
+
+export const JumpToLastMessageButton: FC<JumpToLastMessageButtonProps> = ({
```
You don't need to pass FC<..> here, You have already typed on line 379 :)

svitovyda replied:
done

przemvs commented on src/script/components/MessagesList/MessageList.tsx:
```diff
@@ -245,78 +252,158 @@ export const MessagesList: FC<MessagesListParams> = ({
+      onClick={onGoToLastMessage}
+      css={jumpToLastMessageButtonStyles(mdBreakpoint)}
+    >
+      <ChevronIcon css={{rotate: '90deg', height: 16, width: 16, path: {fill: '#0667C8'}}} />
```
Please use variables for fill :)

svitovyda replied:
where should I add it? this is a completely new color AFAIK, we don't have it in our ui-kit

V-Gira replied:
Let me check, a new color is always a red flag

V-Gira replied:
```suggestion
      <ChevronIcon css={{rotate: '90deg', height: 16, width: 16, path: {fill: 'var(--accent-color)'}}} />
```

svitovyda replied:
done

przemvs commented on src/script/entity/Conversation.ts:
```diff
@@ -82,6 +82,10 @@ enum TIMESTAMP_TYPE {
 
+export const isLastReceivedMessage = (messageEntity: Message, conversationEntity: Conversation): boolean => {
+  return messageEntity.timestamp() >= conversationEntity.last_event_timestamp();
+};
```
Why not to use hasLastReceivedMessageLoaded, You have declared it in this Class ;)

svitovyda replied:
because in this case, I need to identify if the given (rendered) message is the last message, to add to it `onVisible` and `onVisibilityLost`. but in many cases it can be replaced with `hasLastReceivedMessageLoaded` - I'm planning to do the refactoring of it later, within the bug that I discovered. here I just wanted to move it out of React component, as a first step.

przemvs replied:
So please move this from Class :) We have some utils/helpers for cases like this one :)

svitovyda replied:
done. but I think I will soon move it back as a method :)

Commit:
Fix the chewron color in jump to last message button;
Fix responsive positioning of the button;

Commit:
Move JumpToLastMessageButton to separate file;
Move isLastReceivedMessage to utils;
Move all CSS for button into the styles file;

atomrc commented on src/script/components/MessagesList/MessageList.styles.ts:
```diff
@@ -0,0 +1,41 @@
+  borderRadius: '100%',
+  bottom: '56px',
+
+  '@media (max-width: 768px)': {
+    bottom: '100px',
+  },
```
@V-Gira I see this value being copied a few times in the app. Not a topic for now, but we should have a single place where the different breakpoints are defined so we avoid hardcoding breakpoints like this (and when possible those breakpoints should not be used in favor of just relying on the natural DOM flow).



I did try to use the DOM flow to have the button placing naturally, but turns out it's not really possible to have a button that is fixed in a scrolling container sadly. 



Let's try to optimize this in a followup PR

svitovyda replied:
I also see that we have lots of hardcoded breakpoints in the code, and they are different. We should define key breakpoints, and then each component should use them and introduce new ones as `theme.breakpoints.tablet - COMPONENT_CUSTOM_WIDTH_CONSTANT` etc.
